### PR TITLE
Refactor: freeze requirements, extract functionality from context.py

### PR DESCRIPTION
**Pull request by wsh**
This change:
- makes ndb's requirements more explicit through the use of the pip standard `requirements.txt`
- extracts `AutoBatcher` from `context.py` into `autobatcher.py`
- extracts memcache-related functionality from `Context` into `MemcacheClient` in `memcache_client.py`, while preserving `Context`'s memcache API

### CODE REVIEW

File: ndb/memcache_client.py [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  def memcache_incr(self, key, delta=1, initial_value=None, namespace=None,
+                    deadline=None):
+    if not isinstance(key, basestring):
+      raise TypeError('key must be a string; received %r' % key)
+    if not isinstance(delta, (int, long)):
+      raise TypeError('delta must be a number; received %r' % delta)
+    if initial_value is not None and not isinstance(initial_value, (int, long)):
+      raise TypeError('initial_value must be a number or None; received %r' %
+                      initial_value)
+    if namespace is None:
```
Comment (pcostell) [inline ndb/memcache_client.py]: I'd consider moving the namespace out of the client.

File: ndb/memcache_client.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+from .google_imports import namespace_manager
+from .google_imports import memcache
+
+from . import autobatcher
+from . import tasklets
+
+class MemcacheClient(object):
+  # NOTE: The default memcache prefix is altered if an incompatible change is
+  # required. Remember to check release notes when using a custom prefix.
+  _memcache_prefix = 'NDB9:'  # TODO: Might make this configurable
```
Comment (pcostell) [inline ndb/memcache_client.py]: It seems odd that the prefix is defined, but afaict, not actually used in this file.

Comment (wsh) [reply ndb/memcache_client.py]: Good point. It's only used in `Context`. Remvoed.

File: ndb/memcache_client.py (hunk 1 above)
Comment (wsh) [reply ndb/memcache_client.py]: Yes, absolutely. Maybe as a follow-on change to this? :)

Comment (pcostell) [reply ndb/memcache_client.py]: SGTM

File: ndb/memcache_client.py (hunk 2 above)
Comment (pcostell) [reply ndb/memcache_client.py]: I don't think this PR has picked up that change.

Comment (wsh) [reply ndb/memcache_client.py]: Sorry about that. Wrong branch. >_>

**Pull request merged**